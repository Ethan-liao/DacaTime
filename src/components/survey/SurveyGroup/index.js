import React from 'react';
import PropTypes from 'prop-types';
import R from 'ramda';
import { connect } from 'react-redux';
import TextFieldQuestion from '../TextFieldQuestion';
import MultipleChoiceQuestion from '../MultipleChoiceQuestion';
import MultipleChoiceDropdown from '../MultipleChoiceDropdown';
import MultipleChoiceCheckbox from '../MultipleChoiceCheckbox';
import SeekLegalGuidanceDialog from '../LegalGuidance';
import Button from '../../basic/Button';

const createQuestionElem = props =>
  R.cond([
    [
      R.pathEq(['questionType'], 'redFlag'),
      obj =>
        <SeekLegalGuidanceDialog
          show={props.value && props.value[obj.name] === props.redFlagAnswer}
          key={Math.random()}
          updateKeyPath={[obj.name]}
        />,
    ],
    [
      R.pathEq(['questionType'], 'text'),
      obj =>
        <TextFieldQuestion
          value={props.value && props.value[obj.name]}
          key={obj.questionId}
          {...obj}
        />,
    ],
    [
      R.pathEq(['questionType'], 'mcRadio'),
      obj =>
        <MultipleChoiceQuestion
          value={props.value && props.value[obj.name]}
          key={obj.questionId}
          {...obj}
        />,
    ],
    [
      R.pathEq(['questionType'], 'mcDropdown'),
      obj =>
        <MultipleChoiceDropdown
          value={props.value && props.value[obj.name]}
          key={obj.questionId}
          {...obj}
        />,
    ],
    [
      R.pathEq(['questionType'], 'mcCheckbox'),
      obj =>
        <MultipleChoiceCheckbox
          value={props.value && props.value[obj.name]}
          key={obj.questionId}
          {...obj}
        />,
    ],
    [R.T, _ => <h1>Unrecognized question type</h1>],
  ]);

export const SurveyGroup = props => {
  const questionElems = R.map(
    createQuestionElem(props),
    R.sortBy(R.prop('questionId'), props.questions),
  );
  const previousLocation = `/application/${props.previous || ''}`;
  const nextLocation = `/application/${props.next || ''}`;
  return (
    <div className="dt-survey-group-inner-container">
      {questionElems}
      <div className="dt-survey-group-button-container">
        <Button to={previousLocation} disabled={!props.previous}>
          Previous
        </Button>

        <Button to={nextLocation} disabled={!props.next}>
          Next
        </Button>
      </div>
    </div>
  );
};
SurveyGroup.propTypes = {
  id: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(
    PropTypes.shape({
      questionType: PropTypes.oneOf([
        'mcRadio',
        'mcDropdown',
        'mcCheckbox',
        'text',
        'redFlag',
      ]),
      questionId: PropTypes.number.isRequired,
      prompt: PropTypes.string.isRequired,
      explanation: PropTypes.string,
      fields: PropTypes.arrayOf(
        PropTypes.shape({
          label: PropTypes.string.isRequired,
          name: PropTypes.string.isRequired,
          inputType: PropTypes.string,
        }),
      ).isRequired,
    }),
  ).isRequired,
};

const mapStateToProps = state => ({
  value: state.form,
});

export const component = SurveyGroup;

export default connect(mapStateToProps)(component);
