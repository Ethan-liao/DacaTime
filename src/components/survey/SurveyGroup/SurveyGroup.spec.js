import React from 'react';
import SurveyGroup from './index';
import renderer from 'react-test-renderer';
import { surveyGroup } from '../../../util/testData';
import reducer from '../../../reducers';
import { setup } from '../../../util/testUtils';

const props = {
  previous: null,
  next: null,
  questions: surveyGroup.questions,
  id: surveyGroup.id,
};
const { component, store } = setup(SurveyGroup, reducer, props);

it('renders without crashing', () => {
  renderer.create(component);
});

it('renders correctly', () => {
  const renderedComponent = renderer.create(component);
  let tree = renderedComponent.toJSON();

  expect(tree).toMatchSnapshot();
});
