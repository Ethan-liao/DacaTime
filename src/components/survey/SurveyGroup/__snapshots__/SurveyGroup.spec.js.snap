exports[`test renders correctly 1`] = `
<div
  className="dt-survey-group-inner-container">
  <div
    className="dt-panel dt-question-container">
    <div
      className="dt-question-header">
      Were you under the age of 31 as of June 15, 2012?
    </div>
    <div
      className="dt-question-container-horizontal"
      onChange={[Function]}
      value="">
      <div
        className="dt-radio-container">
        <input
          checked={false}
          className="dt-radio-input"
          hidden={true}
          id="radio-1-yes"
          name="radio-1-yes"
          onChange={[Function]}
          type="radio"
          value="yes" />
        <label
          className="dt-radio-label"
          htmlFor="radio-1-yes">
          Yes
        </label>
        <div
          className="dt-radio-button" />
      </div>
      <div
        className="dt-radio-container">
        <input
          checked={false}
          className="dt-radio-input"
          hidden={true}
          id="radio-1-no"
          name="radio-1-no"
          onChange={[Function]}
          type="radio"
          value="no" />
        <label
          className="dt-radio-label"
          htmlFor="radio-1-no">
          No
        </label>
        <div
          className="dt-radio-button" />
      </div>
    </div>
  </div>
  <div
    className="dt-panel dt-question-container">
    <div
      className="dt-question-header">
      When does your DACA status expire?
    </div>
    <div>
      <p
        className="control-label">
        MM/DD/YYYY
      </p>
      <input
        className="survey-input"
        list={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        pattern="^\\d{2}\\/\\d{2}\\/\\d{4}$"
        placeholder="MM/DD/YYYY"
        type="text"
        value="" />
    </div>
  </div>
  <div
    className="dt-panel dt-question-container">
    <div
      className="dt-question-header">
      What is your current marital status?
    </div>
    <select
      className="survey-select"
      onChange={[Function]}
      value="">
      <option
        value="">
        Select a choice...
      </option>
      <option
        value="married">
        Married
      </option>
      <option
        value="widowed">
        Widowed
      </option>
      <option
        value="single">
        Single
      </option>
      <option
        value="divorced">
        Divorced
      </option>
    </select>
  </div>
  <div
    className="dt-panel dt-question-container">
    <div
      className="dt-question-header">
      Have you been convicted of a felony?
    </div>
    <div
      className="dt-question-explanation">
      <a
        onClick={[Function]}>
        What does this mean?
      </a>
    </div>
    <div
      className="dt-question-container-horizontal"
      onChange={[Function]}
      value="">
      <div
        className="dt-radio-container">
        <input
          checked={false}
          className="dt-radio-input"
          hidden={true}
          id="radio-4-yes"
          name="radio-4-yes"
          onChange={[Function]}
          type="radio"
          value="yes" />
        <label
          className="dt-radio-label"
          htmlFor="radio-4-yes">
          Yes
        </label>
        <div
          className="dt-radio-button" />
      </div>
      <div
        className="dt-radio-container">
        <input
          checked={false}
          className="dt-radio-input"
          hidden={true}
          id="radio-4-no"
          name="radio-4-no"
          onChange={[Function]}
          type="radio"
          value="no" />
        <label
          className="dt-radio-label"
          htmlFor="radio-4-no">
          No
        </label>
        <div
          className="dt-radio-button" />
      </div>
    </div>
    <div />
  </div>
  <div
    className="dt-survey-group-button-container">
    <div
      className="dt-button  dt-button-disabled">
      <div
        className="dt-button-text">
        Previous
      </div>
    </div>
    <div
      className="dt-button  dt-button-disabled">
      <div
        className="dt-button-text">
        Next
      </div>
    </div>
  </div>
</div>
`;
